src/lnd.rs

```diff
@@ -173,6 +173,125 @@ impl LndClient {
         Ok(())
     }
 
+    /// Creates a fully populated PSBT with 1 output to the specified address of the
+    /// specified amount. Coin selection is performed automatically. After either verifying the
+    /// inputs, all input UTXOs are locked with an internal LND app ID.
+    ///
+    /// NOTE: If this method returns without an error, it is the caller's responsibility to either
+    /// spend the locked UTXOs (by finalizing and then publishing the transaction) or to
+    /// unlock/release the locked UTXOs in case of an error on the caller's side.
+    pub async fn fund_original_psbt(
+        &self,
+        address: &bitcoin::Address,
+        amount: bitcoin::Amount,
+    ) -> Result<(PartiallySignedTransaction, Vec<UtxoLease>), LndError> {
+        use tonic_lnd::walletrpc::fund_psbt_request::{Fees, Template};
+        use tonic_lnd::walletrpc::{FundPsbtRequest, TxTemplate};
+
+        let client = &mut *self.0.lock().await;
+        let client = client.wallet();
+
+        let mut outputs = std::collections::HashMap::with_capacity(1);
+        outputs.insert(address.to_string(), amount.as_sat());
+        let tx_template = TxTemplate { outputs, ..Default::default() };
+        let template = Some(Template::Raw(tx_template));
+        let fees = Some(Fees::TargetConf(2));
+        let fund_psbt = FundPsbtRequest { template, fees, ..Default::default() };
+
+        let response = client.fund_psbt(fund_psbt).await?;
+        let stream = response.get_ref();
+        let raw_psbt = stream.funded_psbt.as_slice();
+        let funded_psbt =
+            PartiallySignedTransaction::consensus_decode(raw_psbt).map_err(LndError::BadPsbt)?;
+        Ok((funded_psbt, stream.locked_utxos.clone()))
+    }
+
+    /// Expects a partial transaction with all inputs and outputs fully declared and
+    /// tries to sign all unsigned inputs that have all required fields (UTXO information, BIP32
+    /// derivation information, witness or sig scripts) set. If no error is returned, the PSBT is
+    /// ready to be given to the next signer or to be finalized if lnd was the last signer.
+    ///
+    /// NOTE: This RPC only signs inputs (and only those it can sign), it does not perform any
+    /// other tasks (such as coin selection, UTXO locking or input/output/fee value validation,
+    /// PSBT finalization). Any input that is incomplete will be skipped.
+    pub async fn sign_psbt(
+        &self,
+        funded_psbt: PartiallySignedTransaction,
+    ) -> Result<PartiallySignedTransaction, LndError> {
+        let client = &mut *self.0.lock().await;
+        let client = client.wallet();
+
+        let funded_psbt = bitcoin::consensus::serialize(&funded_psbt);
+        let req = tonic_lnd::walletrpc::SignPsbtRequest { funded_psbt, ..Default::default() };
+        let res = client.sign_psbt(req).await?;
```

Comment:
Just curious, is this "tries to sign all unsigned inputs that have all required fields" logic happening in LND or one of the rust libraries?

Reply:
LND signs as many PSBT inputs as it can with the keys in its keychain. We pass the funded (i.e. inputs specified) PSBT to LND over gRPC client here and await a response with newly-signed inputs.

Reply:
I am wondering if there are any security risks here, if I know an output controlled by LND, can I maliciously include it in the PSBT with an output to myself and this function will sign it without considering where it is going?

I'll have a look for where we check the inputs and outputs are expected

Reply:
The `payjoin` crate makes this check when the proposal is processed. `sign_psbt` here is only used for signing the Original PSBT

https://github.com/chaincase-app/payjoin/blob/358e53bdb4b9e533422a9cb06263f48a99e2785d/payjoin/src/sender/mod.rs#L236-L249